Fix nil pointer dereference in metrics value calculator.

humblec opened the pull request:
Fix nil pointer dereference in metrics value calculator.
    
The metrics structure passed to volume stat calculator
can contain real stats on subset of metrics fields. For example,
the metrics structure filled by a CSI driver can have
either INODES or BYTES filled, IOW it a valid return.
In such cases the volume stat calculator panic with below
trace:
    
```    0516 21:36:19.013143   14452 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
    /usr/local/go/src/runtime/panic.go:522
    /usr/local/go/src/runtime/panic.go:82
    /usr/local/go/src/runtime/signal_unix.go:390
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource/quantity.go:697
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource/quantity.go:685
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/volume_stat_calculator.go:144
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/volume_stat_calculator.go:125
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/volume_stat_calculator.go:65
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
    /home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
    /usr/local/go/src/runtime/asm_amd64.s:1337
```

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

humblec committed:
Fix nil pointer dereference in metrics value calculator.

The metrics structure passed to volume stat calculator
can contain real stats on subset of metrics fields. For example,
the metrics structure filled by a CSI driver can have
either INODES or BYTES filled, IOW it a valid return.
In such cases the volume stat calculator panic with below
trace:

0516 21:36:19.013143   14452 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource/quantity.go:697
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource/quantity.go:685
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/volume_stat_calculator.go:144
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/volume_stat_calculator.go:125
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/kubelet/server/stats/volume_stat_calculator.go:65
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
/home/hchiramm/gopath/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
/usr/local/go/src/runtime/asm_amd64.s:1337

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

mattjmcnaughton reviewed (changes requested):
Thanks for thinking about this :)

Overall, I think the nil check is a great idea - we just need to ensure that we're at least initializing the variables to a default value. See my comment and the test suite failures for more info.

Also, could you please fill in the information in your pr? Specifically, the why we need this pull request section (i.e. desc of bug you're fixing, the impact of the bug, etc...) Thanks!

humblec commented:
>Thanks for thinking about this :)
yw :)

>Overall, I think the nil check is a great idea - we just need to ensure that we're at least initializing the variables to a default value. See my comment and the test suite failures for more info.
>

Done. Updated the patch set.

>
Also, could you please fill in the information in your pr? Specifically, the why we need this pull request section (i.e. desc of bug you're fixing, the impact of the bug, etc...) Thanks!
>

Updated commit message with the details on how did I find the culprit :)

mattjmcnaughton commented:
/test pull-kubernetes-typecheck

mattjmcnaughton commented:
/test pull-kubernetes-verify

mattjmcnaughton commented:
> > Thanks for thinking about this :)
> > yw :)
> 
> > Overall, I think the nil check is a great idea - we just need to ensure that we're at least initializing the variables to a default value. See my comment and the test suite failures for more info.
> 
> Done. Updated the patch set.
> 
> > 
> 
> Also, could you please fill in the information in your pr? Specifically, the why we need this pull request section (i.e. desc of bug you're fixing, the impact of the bug, etc...) Thanks!
> 
> > 
> 
> Updated commit message with the details on how did I find the culprit :)

Thanks for the updated commit message :)

Unfortunately, looks like there are still typecheck errors (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/78222/pull-kubernetes-typecheck/1131945681234169856). You can run `./hack/verify-typecheck.sh` to test these locally.

Other than getting the test suite to pass, I think this is looking great :)

humblec commented:
>Unfortunately, looks like there are still typecheck errors (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/78222/pull-kubernetes-typecheck/1131945681234169856). >

@mattjmcnaughton  Ah.. my bad, was in hurry to push the change while boarding the flight :), corrected it.  Thanks ! PTAL.

mattjmcnaughton reviewed (approved):
/lgtm

thank you for this fix and for making the updates to get it to pass all the checks :)

humblec commented:
/release-note-none

humblec commented:
/retest

humblec commented:
@mattjmcnaughton Thanks , not sure who can approve this PR :)

humblec commented:
@tallclair may be you can approve this ? :)

mattjmcnaughton commented:
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: _[humblec](https://github.com/kubernetes/kubernetes/pull/78222#)_
> To fully approve this pull request, please assign additional approvers.
> We suggest the following additional approver: **tallclair**
> 
> If they are not already assigned, you can assign the PR to them by writing `/assign @tallclair` in a comment when ready.
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=kubernetes%2Fkubernetes).
> 
> The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)
> 
> Needs approval from an approver in each of these files:
> * **[pkg/kubelet/OWNERS](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/OWNERS)**
> 
> Approvers can indicate their approval by writing `/approve` in a comment
> Approvers can cancel approval by writing `/approve cancel` in a comment

This message from the prow robot suggests an approver :)

/assign @tallclair

humblec commented:
May be @thockin can also help  :) . He has helped in many PRs :)

humblec commented:
/assign @dchen1107

humblec commented:
/assign @Random-Liu

tallclair commented on pkg/kubelet/server/stats/volume_stat_calculator.go:
```diff
@@ -138,12 +138,26 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
+	var available, capacity, used, inodes, inodesFree, inodesUsed uint64
+	if metric.Available != nil {
+		available = uint64(metric.Available.Value())
+	}
```
alternatively, you could make a helper function:
```
func getMetricValue(r *resource.Quantity) uint64 {
  if r != nil {
    return uint64(r.Value())
  }
  return 0
```

tallclair reviewed (commented):
/approve

nikhita commented:
/kind bug

jieyu commented:
Will this be backported to prior supported releases?

gnufied commented:
@humblec can we open a backport request for 1.15 please?

humblec commented:
> @humblec can we open a backport request for 1.15 please?

@gnufied filed https://github.com/kubernetes/kubernetes/pull/80616
